Cognitive Complexity (step 1 out of 3+): name changes

felix91gr opened the pull request:
Following up on #3793 

**Overall checklist:**

1. **Name changes**
2. MVP of functionality
3. Tests

After this PR, we will start working on the implementation itself.

felix91gr commented:
Whoops, Travis failed because I didn't run the automatic naming update script

Manishearth commented:
I think we have a way of registering renames so that the old name still works but warns about the lint being renamed.

Also, we should probably have something similar for the conf option, allowing people to migrate.

bors commented:
:umbrella: The latest upstream changes (presumably #3790) made this pull request unmergeable. Please resolve the merge conflicts.

felix91gr commented:
That seems to be the case. The [stability guarantees](https://github.com/rust-lang/rfcs/blob/master/text/2476-clippy-uno.md#stability-guarantees) postulated for Clippy mention that a PR may never quite "remove" a Lint, only deprecate it and warn about its subsequent usage.

Might that be the cause of the error?

Should we be deprecating Cyclomatic Complexity and offering to use Cognitive Complexity (as an entirely different lint) instead?

(CC @oli)

felix91gr commented:
That seems to be the case. The [stability guarantees](https://github.com/rust-lang/rfcs/blob/master/text/2476-clippy-uno.md#stability-guarantees) postulated for Clippy mention that a PR may never quite "remove" a Lint, only deprecate it and warn about its subsequent usage.

Might that be the cause of the error?

Should we be deprecating Cyclomatic Complexity and offering to use Cognitive Complexity (as an entirely different lint) instead?

(CC @oli-obk)

mikerite commented:
Take a look at #3554 if you want an example of renaming a lint.

felix91gr commented:
Thanks @mikerite! (I'll continue tomorrow tho, now it's pretty late)

I think it's great that there's a mechanism in place for renaming lints 😊

oli-obk commented on clippy_lints/src/utils/conf.rs:
```diff
@@ -110,8 +110,8 @@ macro_rules! define_Conf {
-    /// Lint: CYCLOMATIC_COMPLEXITY. The maximum cyclomatic complexity a function can have
-    (cyclomatic_complexity_threshold, "cyclomatic_complexity_threshold", 25 => u64),
+    /// Lint: COGNITIVE_COMPLEXITY. The maximum cognitive complexity a function can have
+    (cognitive_complexity_threshold, "cognitive_complexity_threshold", 25 => u64),
```
Can we somehow still accept the old name in config files or at least not hard error about it?

felix91gr replied:
I think that if I implement a formal name change like it was done in #3554, Clippy will tell you the name was changed.
More than that I don't know how to do, but that will be better than a hard error.

flip1995 replied:
I don't think this will work for the `clippy.toml` out of the box. But we don't have a mechanism in place for that yet.

There should be an unknown error for `cyclomatic_complexity`. Maybe this error can be extended to suggest the new name.
https://github.com/rust-lang/rust-clippy/blob/2141ebf69581ab3865ea8c83545e1941dd4d38af/tests/ui-toml/toml_unknown_key/conf_unknown_key.stderr#L1-L3

Or a new error type can be added here (this error is generated by the `utils::conf::read` function):
https://github.com/rust-lang/rust-clippy/blob/db13e6fa34707c62c5d87aa057813a40c67e29c4/clippy_lints/src/utils/conf.rs#L39-L44

felix91gr replied:
I think the key to change what error is generated lies in one of these two sections?

https://github.com/rust-lang/rust-clippy/blob/db13e6fa34707c62c5d87aa057813a40c67e29c4/clippy_lints/src/utils/conf.rs#L72-L79

https://github.com/rust-lang/rust-clippy/blob/db13e6fa34707c62c5d87aa057813a40c67e29c4/clippy_lints/src/utils/conf.rs#L110-L153

I'd like the error to match against `cyclomatic_complexity_threshold`, hopefully looking up in a list of renamed parameters, and if finding that, throwing a nicer "renaming error" rather than just the "generic name-not-found error"

oli-obk replied:
I have a slightly stupid idea/hack: readd the `(cyclomatic_complexity_threshold, "cyclomatic_complexity_threshold", 16544198613561 => u64),`

Note the overly specific `16544198613561`.

Then, have the code that reads the `cognitive_complexity_threshold` also read the `cyclomatic_complexity_threshold` and warn about an unused and outdated toml variabl if the value isn't `16544198613561`.

felix91gr commented:
This is strange. Github says there are conflicts with the `cyclomatic_complexity.rs` file :thinking: 

![](https://screenshotscdn.firefoxusercontent.com/images/baa2284b-a26d-4e6c-a328-2c303430ee94.png)

felix91gr commented:
Nvm, I think I solved the conflicts!

felix91gr commented:
Yay! All the checks have passed. Finally :sweat_smile: 
Now I think I understand a lot more about how Clippy is tested, tho, so for me it was very much worth the while.

felix91gr commented:
To summarize, the PR now contains these changes:
* Renaming of Cyclomatic Complexity to Cognitive Complexity
* "Formal Renaming", i.e., Clippy now knows and can warn you about the deprecation of the `cyclomatic_complexity` lint if you're using it inside your code.
* Updated the test for Formal Renaming so that this renaming is checked for in the future.
* Updated auto-generated files with new names.
* Updated README.md example of `cognitive_complexity_threshold` configuration.

What's blocking this PR:
* How to tell the user about the renaming, if they are using the (now deprecated) `cyclomatic_complexity_threshold` value in their `Clippy.toml` file.

felix91gr commented:
Reading further into the `define_Conf` macro, I'm more and more certain that the key to putting "good errors for renamed values" in the handling of `clippy.toml` files, lies right there. But I got nothin': I'm barely able to interpret a `macro_rules!` macro and this one's beyond my league. Pls help :sweat_smile: .

bors commented:
:umbrella: The latest upstream changes (presumably #3814) made this pull request unmergeable. Please resolve the merge conflicts.

felix91gr commented:
Alright, I think I've made it tell the user about the usage of the deprecated field `cyclomatic_complexity_threshold`, suggesting using the `cognitive_complexity_threshold` field instead. I've also added a test for it.

felix91gr commented:
CI gives an odd error. Maybe I just missed an upstream change once again 🤔

Anyone got a clue?

detrumi commented:
> CI gives an odd error. Maybe I just missed an upstream change once again

Yes, you'll probably want to rebase so you get the latest fixes

felix91gr commented on clippy_lints/src/utils/conf.rs:
```diff
@@ -227,22 +229,22 @@ pub fn read(path: Option<&path::Path>) -> (Conf, Vec<Error>) {
-            } else {
-                errors.push(Error::Toml(err_as_str));
+
+            let cyc_field: u64 = (toml as Conf).cyclomatic_complexity_threshold;
```
I added this because the compiler didn't let me do `toml.(...)` directly, apparently it needed a type annotation? The way I just did it is probably not what it needs, tho. If this commit works, then I'll fix the type annotation later.

felix91gr replied:
Update: did not work.

felix91gr commented:
The changes seemingly worked. Now I'm adding a test to check that the `conf.rs` file doesn't throw a false negative on a correct `clippy.toml` file.

felix91gr commented:
Okay, all checks have passed and now `conf.rs` gives the appropiate error if and only if the user has an otherwise valid `clippy.toml` file with the `cyclomatic-complexity-threshold` field in it.

flip1995 commented:
I have one more case that should be tested:
https://github.com/rust-lang/rust-clippy/blob/da7aebc342b7dae444912b98dedbab22bf2a224c/tests/ui/cyclomatic_complexity.rs#L90

It is currently possible to set the cyclomatic_complexity with an attribute. What is the output of the compiler if you sill use this attribute? Could you add a test for this?

felix91gr commented:
Ohh, that's right! I forgot you could set it in-code again after re-adding the CyC to the macro. Hmm 🤔 I'll have to study the conf part a little bit more. 

Do you happen to know where is the in-file setting read?

Maybe a stub implementation inside of `clippy-lint/` must be present for the setting not to crash Clippy.

felix91gr commented:
> It is currently possible to set the cyclomatic_complexity with an attribute. What is the output of the compiler if you sill use this attribute? Could you add a test for this?

It is apparently invisible. I made a test like this:

```rust
#[clippy::cyclomatic_complexity = "0"]
fn main() {
}
```

And the compiler and test said nothing (not even in `stderr`). It seems to only react when you're using the top-level configuration. This, for example:

```rust
#![warn(clippy::cyclomatic_complexity)]
#[clippy::cyclomatic_complexity = "0"]
fn main() {
}
```

Throws something like this:

```
error: lint `clippy::cyclomatic_complexity` has been renamed to `clippy::cognitive_complexity`
  --> $DIR/rename.rs:2:9
   |
LL | #![warn(clippy::cyclomatic_complexity)]
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: use the new name: `clippy::cognitive_complexity`
   |
   = note: `-D renamed-and-removed-lints` implied by `-D warnings`
```

Is that okay? Is the value-setter alone enough to trigger the lint? It should crash, shouldn't it? Then I think it didn't trigger at all, because I didn't see it crash. Maybe the setting is invisible to Clippy if you're not using the `#![warn(clippy::cyclomatic_complexity)]` line :thinking:

flip1995 commented:
Oh yeah, that's kind of the behavior I expected. From the [RFC](https://github.com/rust-lang/rfcs/blob/master/text/2103-tool-attributes.md#activation-and-unused-attibuteslints):

> A tool that has an active name must check for unused lints/attibutes. For example, if rustfmt becomes active for attributes, and only recognises rustfmt::skip, it must produce a warning if a user uses rustfmt::foo in their code.

We already implemented a lint for unknown Clippy lints (#3161). This should also be done for Clippy attributes. I try to get to this today. I will notify you if I'm done.

flip1995 commented:
#3830 will give you the ability to add the new attribute `cognitive_complexity` to the list of `BUILTIN_ATTRIBUTES` and deprecate the old attribute.

felix91gr commented:
Your PR got merged, that's awesome!

Btw @flip1995, what kinds of attributes are there? "Builtin" is one, what are the other ones?

flip1995 commented:
I don't think there are other types of attributes in Clippy. I just copied the naming from rustc https://github.com/rust-lang/rust/blob/1999a2288123173b2e487865c9a04386173025f7/src/libsyntax/feature_gate.rs#L828

In rustc there are for example "tool attributes", which are provided by tools like Clippy and rustfmt. Also there are "custom attributes".

oli-obk commented on clippy_lints/src/utils/attrs.rs:
```diff
@@ -15,7 +15,10 @@ pub enum DeprecationStatus {
+    (
+        "cyclomatic_complexity",
+        DeprecationStatus::Replaced("cognitive_complexity"),
+    ),
```
So if I understand this list correctly, you'll need to add an entry for `cognitive_complexity` with `DeprecationStatus::None`, otherwise you'll get complaints about an unknown attribute when you start using `#[clippy::cognitive_complexity(..)]`

felix91gr replied:
Ahh! I see. Thank you :D

felix91gr commented:
(Unrelated to the attributes talk,) I'm getting another compilation error, but this time I'm almost certain I'm running the `master` toolchain.

<details><summary>This is the error</summary>

![image](https://user-images.githubusercontent.com/11747623/53649424-df086480-3c20-11e9-92c9-dc1260e80530.png)

</details>

And these are the commands I ran before I got that result:

```
./setup-toolchain.sh
./ci/base-tests.sh
```

flip1995 commented:
The fix for this error is on it's way: #3823

felix91gr commented:
So to recap, the PR has now:

* Error for use in config files, with the error telling the user that they should change it to "cognitive-complexity". **Tested**
* Error for use in .rs files (as top level `#![warn(clippy::cyclomatic_complexity)]`) telling the user that the lint has been renamed. **Tested**
* Error for use in .rs files as builtin attribute, telling the user that the attribute has been renamed. **Tested**

I think all the user-facing errors have been implemented. Anything else I might be missing? @oli-obk

oli-obk commented on clippy_lints/src/utils/conf.rs:
```diff
@@ -110,8 +110,10 @@ macro_rules! define_Conf {
+    /// Lint: COGNITIVE_COMPLEXITY. The maximum cognitive complexity a function can have
+    (cognitive_complexity_threshold, "cognitive_complexity_threshold", 25 => u64),
+    /// DEPRECATED LINT: CYCLOMATIC_COMPLEXITY. Use the Cognitive Complexity lint instead.
+    (cyclomatic_complexity_threshold, "cyclomatic_complexity_threshold", 11_111_011_111 => u64),
```
Ah, now I understand your confusion with what I was suggesting. My suggestion with using the field directly made no sense, since the fields aren't an option, but defaulted. Rereading the macro expanded code, I think we can make the default value `None` and the type `Option<u64>`. Then we don't need magic numbers but you can check `is_some()` instead.

Sorry about that.

felix91gr replied:
Ah, I see now. Yes, I didn't know how to check it otherwise. I'm testing now a couple of changes that should (I think) let the same code work, but by defaulting to a `None` value instead of a magic number. The one thing I'm wondering is if using an `Option<u64>` will be the same for the toml parsing code as it was using a `u64` directly. We'll see :) 
Thank you for clarifying that ^^

oli-obk commented:
Ok, this looks good to me now. Can you just squash your commits then I'll send it off to bors

felix91gr commented:
Sure!

felix91gr commented:
Okay after the squashing I had solved an old merge conflict leaving behind the wrong line, twice :sweat_smile: . I've been really sleepy today >.<

But I double checked it this time. The rebase should work! (still me -> :crossed_fingers:)

bors commented:
:umbrella: The latest upstream changes (presumably #3845) made this pull request unmergeable. Please resolve the merge conflicts.

felix91gr commented:
Resolved upstream conflict and rebased again.

phansch commented on tests/ui/rename.rs:
```diff
@@ -1,4 +1,5 @@
 #![allow(stutter)]
+#![warn(clippy::cyclomatic_complexity)]
```
Could you make these two (+ `clippy::stutter`) `allow`? Otherwise the evaluation will stop at the first `warn` due to `-D warnings` on CI. This should result in `tests/ui/rename.stderr` including output for `clippy::stutter` again.

felix91gr replied:
Are you sure about that? [This](https://github.com/rust-lang/rust-clippy/blob/ddc718087ffeb7b932da87c26fd4e75ec6a9ac02/tests/ui/rename.stderr) is the `stderr` I get with the current file, and there Clippy is warning about stutter.

Maybe I haven't understood what you meant :thinking:

phansch replied:
hmm, I think I misinterpreted the stderr. It's indeed there, sorry about that =)

phansch commented:
@bors r=oli-obk

bors commented:
:pushpin: Commit ddc718087ffeb7b932da87c26fd4e75ec6a9ac02 has been approved by `oli-obk`

<!-- @bors r=oli-obk ddc718087ffeb7b932da87c26fd4e75ec6a9ac02 -->

bors commented:
:hourglass: Testing commit ddc718087ffeb7b932da87c26fd4e75ec6a9ac02 with merge 00baf7a1118f4b8332a47f68e8bebbac963a5dcb...
<!-- homu: {"type":"BuildStarted","head_sha":"ddc718087ffeb7b932da87c26fd4e75ec6a9ac02","merge_sha":"00baf7a1118f4b8332a47f68e8bebbac963a5dcb"} -->

bors commented:
:sunny: Test successful - [checks-travis](https://travis-ci.com/rust-lang/rust-clippy/builds/103366777), [status-appveyor](https://ci.appveyor.com/project/rust-lang-libs/rust-clippy/builds/22867366)
Approved by: oli-obk
Pushing 00baf7a1118f4b8332a47f68e8bebbac963a5dcb to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"oli-obk","base_ref":"master","builders":{"status-appveyor":"https://ci.appveyor.com/project/rust-lang-libs/rust-clippy/builds/22867366","checks-travis":"https://travis-ci.com/rust-lang/rust-clippy/builds/103366777"},"merge_sha":"00baf7a1118f4b8332a47f68e8bebbac963a5dcb"} -->

felix91gr commented:
⛵️
